rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+`EagerTensor`s as outputs. If called in a graph context (i.e., in
+`tf-function`), it'll construct a new symbolic op by consuming inbound symbolic
+Tensors. Our proposal captures both cases with a single API.
+
+The proposed API achieves Capabilities A and B listed above, due to the fact
+that the callback(s) registered with the API will be invoked for both eager op
+execution and graph op creation.
+
+Capability C is also met by this API, due to the fact that the callbacks are
+invoked not only during the eager execution of ops, but also the execution
+of FuncGraphs. In this paradigm, when a FuncGraph has been constructed and
+undergoes execution by TF2, it is simply treated as a special type of op.
+
+The example code snippet below shows how Capability A is met, i.e., how eager
+execution of ops is intercepted by the callback mechanism.
+
+```python
+with tf.debugging.op_callback(my_callback):
+  x = tf.constant(3.0)
+
+  y = tf.math.log(1 + x)
+  # ↑ During the execution of the line above, `my_callback` is invoked twice:
+  #   1. With the `op_type` arg being 'AddV2' and the `outputs` arg being an
+  #      EagerTensor with value 4.0.
+  #   2. With the `op_type` arg being 'Log' and the `outputs` arg being an
+  #      EagerTensor with value 1.386 (≈log(4.0)).
+```
+
+The code below illustrates how Capability B is met, i.e., how creation of
+ops inside a user-defined Python function decorated by `tf-function`:
+
+```python
+with tf.debugging.op_callback(my_callback):
+
+  @def_function
+  def log_1plusp(p):
+    return tf.math.log(1 + p)
+
+  x = tf.constant(3.0)
+
+  y = log_1plusp(x)
+  # ↑ During the execution of the line above, `my_callback` is invoked *three*
+  #   times. The first two invocations happen during AutoGraph's transformation
+  #   of the function `log_1plusp` into a FuncGraph:
+  #   1. With the `op_type` arg being 'AddV2' and the `outputs` arg being the
+  #      symbolic Tensor output by by AddV2 op.
+  #   2. With the `op_type` arg being 'Log' and the `outputs` arg being the
+  #      symbolic Tensor output by by Log op.
+  #   (In reality, tf.function and AutoGraph may create additional ops such as
+  #    constant ops for Python constants present in the Python function and
+  #    Identity ops to marshal the FuncGraph's input and output values. Those
+  #    extra ops will be captured by `my_callback` as well.)
+  #
+  #   The third (last) invocation of the callback is due to the eager execution
+  #   of the FuncGraph, with the `op_type` arg being `tf.Graph`, the `op_name`
+  #   arg being something like `_inference_log_1plusp_30`, and the `outputs` arg
+  #   being an EagerTensor of value 1.386 (≈log(4.0)).
+```
+
+The example above dealt with a relatively simple FuncGraph. But the proposed API
+applies to more complex FuncGraphs, including those that involve control flow
+such as if-else and while loops. For instance, see the code example below.
+
+```python
+with tf.debugging.op_callback(my_callback):
```

Comment:
Is it worth considering more generic names given that this operates over non-ops? tf.debugging.callback() ?